### ISSUE TITLE
[HW Accel Support]: Unable to use hwaccel on Intel 12th CPU + Docker container

### ISSUE THREAD

**Issue by t0ny-peng** [blakeblackshear/frigate#3934]
### Describe the problem you are having

I have several cameras that have a main stream and a sub stream, as shown in the following table. Frigate runs inside a docker container in a Proxmox VE instance. The CPU is Intel i3-12100. I tried to enable hardware accelerate by adding `hwaccel_args: -c:v h264_qsv` to the `ffmpeg` field of one camera, but the camera is not able to start.

Possibly the same problem in https://github.com/blakeblackshear/frigate/issues/3170 but need confirmation

|Main stream| Sub stream|
|-|-|
|<img width="552" alt="image" src="https://user-images.githubusercontent.com/4816327/192129746-7b650857-8398-47c0-9bf9-a237cebcd061.png">|<img width="552" alt="image" src="https://user-images.githubusercontent.com/4816327/192129736-2067b719-271d-4cf3-ae64-5a9e120237ad.png">|

I tried to manually run the `ffmpeg` command inside the Docker container with the hardware acceleration flag, and it dumps some error message, attached at the below section. This command, without the hardware acceleration part, can dump the stream into pipe without any problem.

```
ffmpeg -hide_banner -c:v h264_qsv -loglevel warning -avoid_negative_ts make_zero \
  -fflags +genpts+discardcorrupt -rtsp_transport tcp -timeout 5000000 -use_wallclock_as_timestamps 1 \
  -i "rtsp://REDACTED:554/cam/realmonitor?channel=1&subtype=1" -f segment -segment_time 10 \
  -segment_format mp4 -reset_timestamps 1 -strftime 1 -c copy -an /tmp/cache/Backyard-%Y%m%d%H%M%S.mp4 \
  -c copy -f flv rtmp://127.0.0.1/live/Backyard -r 5 -s 1280x720 -f rawvideo -pix_fmt yuv420p pipe:
```

### Version

0.11.0

### Frigate config file

```yaml
mqtt:
  host: REDACTED
  user: bot
  password: "REDACTED"

detectors:
  coral:
    type: edgetpu
    device: usb

detect:
  fps: 5
  enabled: True
  max_disappeared: 25

objects:
  track:
    - person
    - cat

record:
  enabled: True
  retain:
    days: 7
    mode: motion
  events:
    retain:
      default: 14
      mode: active_objects

snapshots:
  enabled: True
  timestamp: True
  bounding_box: True

rtmp:
  enabled: True

cameras:
  Backyard_1:
    ffmpeg:
      inputs:
        # 8M stream
        - path: rtsp://REDACTED:554/cam/realmonitor?channel=1&subtype=0
          roles:
            - record
            - rtmp
        # Sub stream
        - path: rtsp://REDACTED:554/cam/realmonitor?channel=1&subtype=1
          roles:
            - detect
      hwaccel_args: -c:v h264_qsv

    detect:
      width: 640
      height: 480
```

### docker-compose file or Docker CLI command

```yaml
version: "3.9"
services:
  frigate:
    container_name: frigate
    privileged: true
    restart: unless-stopped
    image: blakeblackshear/frigate:0.11.0
    shm_size: "1024mb"
    devices:
      - /dev:/dev
    volumes:
      - /etc/localtime:/etc/localtime:ro
      - /root/docker/frigate/frigate.yml:/config/config.yml:ro
      - NFS_Share:/media/frigate:rw
      - type: tmpfs # Optional: 1GB of memory, reduces SSD/SD Card wear
        target: /tmp/cache
        tmpfs:
          size: 1000000000
    ports:
      - "5000:5000"
      - "1935:1935" # RTMP feeds
    environment:
      FRIGATE_RTSP_PASSWORD: "REDACTED"
```

#### Frigate log

```shell
[2022-09-24 22:51:05] watchdog.Backyard_1            INFO    : No frames received from Backyard_1 in 20 seconds. Exiting ffmpeg...
[2022-09-24 22:51:05] watchdog.Backyard_1            INFO    : Waiting for ffmpeg to exit gracefully...
[2022-09-24 22:51:05] frigate.video                  ERROR   : Backyard_1: Unable to read frames from ffmpeg process.
[2022-09-24 22:51:05] frigate.video                  ERROR   : Backyard_1: ffmpeg process is not running. exiting capture thread...
```

#### ffmpeg ran manually inside the container

```shell


[h264_qsv @ 0x55de82f62100] Error during set display handle
: device failed (-17)
[h264_qsv @ 0x55de82f62100] Error initializing an MFX session
[h264_qsv @ 0x55de82f62100] Error decoding header
Error while decoding stream #0:0: Input/output error
[segment @ 0x55de830e23c0] Timestamps are unset in a packet for stream 0. This is deprecated and will stop working in the future. Fix your code to set the timestamps properly
[flv @ 0x55de82f58cc0] Timestamps are unset in a packet for stream 0. This is deprecated and will stop working in the future. Fix your code to set the timestamps properly
[segment @ 0x55de830e23c0] Non-monotonous DTS in output stream 0:0; previous: 0, current: 0; changing to 1. This may result in incorrect timestamps in the output file.
[h264_qsv @ 0x55de82f62100] Error during QSV decoding.: device failed (-17)
Error while decoding stream #0:0: Input/output error
[h264_qsv @ 0x55de82f62100] Error during QSV decoding.: device failed (-17)
Error while decoding stream #0:0: Input/output error
####### Repeated many times ######
Finishing stream 2:0 without any data written to it.
[swscaler @ 0x55de839bf900] [swscaler @ 0x55de839cc8c0] deprecated pixel format used, make sure you did set range correctly
[swscaler @ 0x55de839bf900] [swscaler @ 0x55de83a06140] deprecated pixel format used, make sure you did set range correctly
[swscaler @ 0x55de839bf900] [swscaler @ 0x55de83a3ea00] deprecated pixel format used, make sure you did set range correctly
[swscaler @ 0x55de839bf900] [swscaler @ 0x55de83a772c0] deprecated pixel format used, make sure you did set range correctly
[swscaler @ 0x55de839bf900] [swscaler @ 0x55de83aafb80] deprecated pixel format used, make sure you did set range correctly
[swscaler @ 0x55de839bf900] [swscaler @ 0x55de83ae8fc0] deprecated pixel format used, make sure you did set range correctly
[swscaler @ 0x55de839bf900] [swscaler @ 0x55de83b20d00] deprecated pixel format used, make sure you did set range correctly
[swscaler @ 0x55de839bf900] [swscaler @ 0x55de83b5a140] deprecated pixel format used, make sure you did set range correctly
[swscaler @ 0x55de839bf900] [swscaler @ 0x55de83b92a00] deprecated pixel format used, make sure you did set range correctly
[flv @ 0x55de82f58cc0] Failed to update header with correct duration.
[flv @ 0x55de82f58cc0] Failed to update header with correct filesize.
```

### FFprobe output from your camera

```shell
ffprobe version n5.1-2-g915ef932a3-20220731 Copyright (c) 2007-2022 the FFmpeg developers
  built with gcc 12.1.0 (crosstool-NG 1.25.0.55_3defb7b)
  configuration: --prefix=/ffbuild/prefix --pkg-config-flags=--static --pkg-config=pkg-config --cross-prefix=x86_64-ffbuild-linux-gnu- --arch=x86_64 --target-os=linux --enable-gpl --enable-version3 --disable-debug --enable-iconv --enable-libxml2 --enable-zlib --enable-libfreetype --enable-libfribidi --enable-gmp --enable-lzma --enable-fontconfig --enable-libvorbis --enable-opencl --enable-libpulse --enable-libvmaf --enable-libxcb --enable-xlib --enable-amf --enable-libaom --enable-libaribb24 --enable-avisynth --enable-libdav1d --enable-libdavs2 --disable-libfdk-aac --enable-ffnvcodec --enable-cuda-llvm --enable-frei0r --enable-libgme --enable-libass --enable-libbluray --enable-libjxl --enable-libmp3lame --enable-libopus --enable-mbedtls --enable-librist --enable-libtheora --enable-libvpx --enable-libwebp --enable-lv2 --enable-libmfx --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-libopenh264 --enable-libopenjpeg --enable-libopenmpt --enable-librav1e --enable-librubberband --disable-schannel --enable-sdl2 --enable-libsoxr --enable-libsrt --enable-libsvtav1 --enable-libtwolame --enable-libuavs3d --enable-libdrm --enable-vaapi --enable-libvidstab --enable-vulkan --enable-libshaderc --enable-libplacebo --enable-libx264 --enable-libx265 --enable-libxavs2 --enable-libxvid --enable-libzimg --enable-libzvbi --extra-cflags=-DLIBTWOLAME_STATIC --extra-cxxflags= --extra-ldflags=-pthread --extra-ldexeflags=-pie --extra-libs='-ldl -lgomp' --extra-version=20220731
  libavutil      57. 28.100 / 57. 28.100
  libavcodec     59. 37.100 / 59. 37.100
  libavformat    59. 27.100 / 59. 27.100
  libavdevice    59.  7.100 / 59.  7.100
  libavfilter     8. 44.100 /  8. 44.100
  libswscale      6.  7.100 /  6.  7.100
  libswresample   4.  7.100 /  4.  7.100
  libpostproc    56.  6.100 / 56.  6.100
Input #0, rtsp, from 'rtsp://REDACTED/cam/realmonitor?channel=1&subtype=1':
  Metadata:
    title           : Media Server
  Duration: N/A, start: 0.400000, bitrate: N/A
  Stream #0:0: Video: h264 (High), yuvj420p(pc, bt709, progressive), 640x480 [SAR 2571:1928 DAR 857:482], 5 fps, 100 tbr, 90k tbn
```

### Operating system

Proxmox

### Install method

Docker Compose

### Network connection

Wired

### Camera make and model

AMCREST IP8M-T2499EW-28MM

**Comment by t0ny-peng** [comment]
A guy mentioned (https://github.com/blakeblackshear/frigate/issues/3170#issuecomment-1194515206) that it's probably because the default Proxmox VE kernel is still on 5.15. I'm going to see what feature requires 5.18

**Comment by NickM-27** [comment]
You're not passing the renderD128 to the container in docker compose so it has nothing to accelerate with

**Comment by t0ny-peng** [comment]
@NickM-27 Thanks for the reply. I'm mounting the entire `/dev` folder

```yaml
    devices:
      - /dev:/dev
```

Inside the container the accelerate device can be found. 
```
root@6c07a7d31f9e:/opt/frigate# cd /dev
root@6c07a7d31f9e:/dev# find . -name "renderD128"
./dri/renderD128
```

**Comment by NickM-27** [comment]
I see, interesting. What is your host OS? 

Can you run vainfo in the container and paste the logs here?

**Comment by t0ny-peng** [comment]
I am using Proxmox 7.2, which is based on Debian. It's probably because the Kernel is still on 5.15. Trying to update now as suggested in the other issue.

```
root@0d6e19ca3ca0:/opt/frigate# uname -a
Linux 0d6e19ca3ca0 5.15.30-2-pve #1 SMP PVE 5.15.30-3 (Fri, 22 Apr 2022 18:08:27 +0200) x86_64 GNU/Linux
```

`vainfo` from the container:
```
root@0d6e19ca3ca0:/opt/frigate# vainfo
error: XDG_RUNTIME_DIR not set in the environment.
error: can't connect to X server!
libva info: VA-API version 1.10.0
libva info: Trying to open /usr/lib/x86_64-linux-gnu/dri/iHD_drv_video.so
libva info: Found init function __vaDriverInit_1_10
libva info: va_openDriver() returns 0
vainfo: VA-API version: 1.10 (libva 2.10.0)
vainfo: Driver version: Intel iHD driver for Intel(R) Gen Graphics - 21.1.1 ()
vainfo: Supported profile and entrypoints
      VAProfileNone                   :	VAEntrypointVideoProc
      VAProfileNone                   :	VAEntrypointStats
      VAProfileMPEG2Simple            :	VAEntrypointVLD
      VAProfileMPEG2Simple            :	VAEntrypointEncSlice
      VAProfileMPEG2Main              :	VAEntrypointVLD
      VAProfileMPEG2Main              :	VAEntrypointEncSlice
      VAProfileH264Main               :	VAEntrypointVLD
      VAProfileH264Main               :	VAEntrypointEncSlice
      VAProfileH264Main               :	VAEntrypointFEI
      VAProfileH264Main               :	VAEntrypointEncSliceLP
      VAProfileH264High               :	VAEntrypointVLD
      VAProfileH264High               :	VAEntrypointEncSlice
      VAProfileH264High               :	VAEntrypointFEI
      VAProfileH264High               :	VAEntrypointEncSliceLP
      VAProfileVC1Simple              :	VAEntrypointVLD
      VAProfileVC1Main                :	VAEntrypointVLD
      VAProfileVC1Advanced            :	VAEntrypointVLD
      VAProfileJPEGBaseline           :	VAEntrypointVLD
      VAProfileJPEGBaseline           :	VAEntrypointEncPicture
      VAProfileH264ConstrainedBaseline:	VAEntrypointVLD
      VAProfileH264ConstrainedBaseline:	VAEntrypointEncSlice
      VAProfileH264ConstrainedBaseline:	VAEntrypointFEI
      VAProfileH264ConstrainedBaseline:	VAEntrypointEncSliceLP
      VAProfileHEVCMain               :	VAEntrypointVLD
      VAProfileHEVCMain               :	VAEntrypointEncSlice
      VAProfileHEVCMain               :	VAEntrypointFEI
      VAProfileHEVCMain               :	VAEntrypointEncSliceLP
      VAProfileHEVCMain10             :	VAEntrypointVLD
      VAProfileHEVCMain10             :	VAEntrypointEncSlice
      VAProfileHEVCMain10             :	VAEntrypointEncSliceLP
      VAProfileVP9Profile0            :	VAEntrypointVLD
      VAProfileVP9Profile0            :	VAEntrypointEncSliceLP
      VAProfileVP9Profile1            :	VAEntrypointVLD
      VAProfileVP9Profile1            :	VAEntrypointEncSliceLP
      VAProfileVP9Profile2            :	VAEntrypointVLD
      VAProfileVP9Profile2            :	VAEntrypointEncSliceLP
      VAProfileVP9Profile3            :	VAEntrypointVLD
      VAProfileVP9Profile3            :	VAEntrypointEncSliceLP
      VAProfileHEVCMain12             :	VAEntrypointVLD
      VAProfileHEVCMain12             :	VAEntrypointEncSlice
      VAProfileHEVCMain422_10         :	VAEntrypointVLD
      VAProfileHEVCMain422_10         :	VAEntrypointEncSlice
      VAProfileHEVCMain422_12         :	VAEntrypointVLD
      VAProfileHEVCMain422_12         :	VAEntrypointEncSlice
      VAProfileHEVCMain444            :	VAEntrypointVLD
      VAProfileHEVCMain444            :	VAEntrypointEncSliceLP
      VAProfileHEVCMain444_10         :	VAEntrypointVLD
      VAProfileHEVCMain444_10         :	VAEntrypointEncSliceLP
      VAProfileHEVCMain444_12         :	VAEntrypointVLD
      VAProfileHEVCSccMain            :	VAEntrypointVLD
      VAProfileHEVCSccMain            :	VAEntrypointEncSliceLP
      VAProfileHEVCSccMain10          :	VAEntrypointVLD
      VAProfileHEVCSccMain10          :	VAEntrypointEncSliceLP
      VAProfileHEVCSccMain444         :	VAEntrypointVLD
      VAProfileHEVCSccMain444         :	VAEntrypointEncSliceLP
      VAProfileAV1Profile0            :	VAEntrypointVLD
      VAProfileHEVCSccMain444_10      :	VAEntrypointVLD
      VAProfileHEVCSccMain444_10      :	VAEntrypointEncSliceLP
```

**Comment by t0ny-peng** [comment]
No luck :( With 5.18.19, the error is still the same. Next step I'll try to install Frigate on the host and then see if it's the problem of docker container.

**Comment by NickM-27** [comment]
No need to install frigate on the host. Just run ffmpeg from the host

**Comment by t0ny-peng** [comment]
Hmm. Even on host this command failed with the following message. I'm wondering if `Driver not found in known nonstandard list, using standard behaviour.` indicates something. 

```
[AVHWDeviceContext @ 0x5557285638c0] Format 0x36313459 -> unknown.
[AVHWDeviceContext @ 0x5557285638c0] VAAPI driver: Intel iHD driver for Intel(R) Gen Graphics - 21.1.1 ().
[AVHWDeviceContext @ 0x5557285638c0] Driver not found in known nonstandard list, using standard behaviour.
[h264_qsv @ 0x55572849a100] Error during set display handle
: device failed (-17)
[h264_qsv @ 0x55572849a100] Error initializing an MFX session
[h264_qsv @ 0x55572849a100] Error initializing the MFX video decoder: invalid handle (-6)
Error while decoding stream #0:0: Invalid argument
cur_dts is invalid st:0 (0) [init:0 i_done:0 finish:0] (this is harmless if it occurs once at the start per stream)
[h264_qsv @ 0x55572849a100] Format nv12 chosen by get_format().
[h264_qsv @ 0x55572849a100] Decoder: output is system memory surface
[h264_qsv @ 0x55572849a100] Format nv12 chosen by get_format().
[h264_qsv @ 0x55572849a100] Decoder: output is system memory surface
[h264_qsv @ 0x55572849a100] Error during QSV decoding.: device failed (-17)
Error while decoding stream #0:0: Input/output error
```

**Comment by NickM-27** [comment]
> Driver not found in known nonstandard list, using standard behaviour.

That means the driver is normal ie not nonstandard

It definitely needs to be working on the host to work in the container.

**Comment by ErroneousBosch** [comment]
I am seeing likely this issue also happening on TrueNAS SCALE 22.12 Bluefin with a 12th gen Alder Lake i915, 5.15.79 kernel. Device shows up in the container in /dev/dri, but ffmpeg keeps crashing giving a -17 device error. The interface just shows green for the camera screens.

**Comment by ErroneousBosch** [comment]
Let me also say that Plex running on the same system has no problems, though it uses its own transcoder based on ffmpeg

**Comment by NickM-27** [comment]
This has already been fixed for 0.12
